File: leetcode/232_Implement_queue_using_stacks/solution.py
class MyQueue:
    def __init__(self):
        self.push_stack = []
        self.pop_stack = []
        self.push_stack_front = None

    def push(self, x: int) -> None:
        if not self.push_stack:
            self.push_stack_front = x

        self.push_stack.append(x)

    def pop(self) -> int:
        if not self.pop_stack:
            while self.push_stack:
                self.pop_stack.append(self.push_stack.pop())

        return self.pop_stack.pop()

    def peek(self) -> int or None:
        if self.pop_stack:
            return self.pop_stack[-1]
        if self.push_stack:
            return self.push_stack_front
        return None

    def empty(self) -> bool:
        return not self.push_stack and not self.pop_stack

    def __repr__(self):
        repr_stack = list(reversed(self.pop_stack)) + self.push_stack
        return str(repr_stack)


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: leetcode/232_Implement_queue_using_stacks/solution.ts
export class MyQueue {
  // pushStack is used to store the push elements
  private pushStack: number[] = [];
  // popStack is used to store those ready to pop elements, when no elements here, pop it from pushStack
  private popStack: number[] = [];
  private frontend: number | null = null;

  public push(x: number): void {
    if (this.pushStack.length === 0) {
      this.frontend = x;
    }
    this.pushStack.push(x);
  }

  public pop(): number | null {
    if (this.popStack.length === 0) {
      while (this.pushStack.length > 0) {
        this.popStack.push(this.pushStack.pop());
      }
    }
    if (this.popStack.length === 0) {
      return null;
    }
    return this.popStack.pop();
  }

  public peek(): number | null {
    if (this.popStack.length > 0) {
      return this.popStack[this.popStack.length - 1];
    }
    if (this.pushStack.length > 0) {
      return this.frontend;
    }
    return null;
  }

  public empty(): boolean {
    return this.pushStack.length === 0 && this.popStack.length === 0;
  }

  public toString(): string {
    const reveredPopStack = [...this.popStack].reverse();
    return [...this.pushStack, ...reveredPopStack].toString();
  }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */
